### PR TITLE
Find references of the base field or include descendant fields

### LINKED ISSUE #9044: [display] Find references vs method overrides
Currently, "find references" requests only finds exact method calls, not including superclass/interface method calls in case of virtual method overriding. So with subclassing/interfaces, you often have to travel through the inheritance chain and do "find references" for the same method several times.

I understand that the current behaviour also might be desirable, so I think this should be somehow configurable. For example, IIRC, IntelliJ provides a selection on whether you want to see exact matches or include all the virtual calls. I'm not sure if VS Code supports this, but it could at least be a setting or some memento-stored switch.

Some code to illustrate the problem:
```haxe
class Main {
	static function main() {
		var b = new B();
		var a:A = b;
		var i:I = b;

		i.f();
		a.f();
		b.f();
	}
}

interface I {
	function f():Void; // only finds i.f()
}

class A implements I {
	public function f() {} // only finds a.f()
}

class B extends A {
	public function new() {}
	override function f() {} // only finds b.f()
}
```

This is quite unfortunate for inheritance-heavy codebases such as the one we have at @innogames :)

### LINKED ISSUE #9139: [display] "Find references" doesn't find interface fields accessed through the class directly
Trying to "find references" on `I.g` will find nothing, while I think it should include the direct usages of the methods of implementor classes.

```haxe
class Main {
	static function main() {
		new C().g();
	}
}

class C implements I {
	public function new() {}
	public function g() {}
}

interface I {
	function g():Void;
}
```

### PR DESCRIPTION
**Pull request by RealyUniqueName**
Closes #9044
Closes #9139 (using `withDescendants` kind; see below)

This PR adds optional `kind` parameter to the `display/references` request:
* `"direct"` - old behavior
* `"withBaseAndDescendants'` - find references to the base field instead of the requested one. Also finds references to all descendants of the base field.
*  `"withDescendants"` - find references to the requested field and references to all descendants of the requested field.

It is possible to use this parameter with command line requests:
```
$ haxe build.hxml --display src/Main.hx@75@usage@withBaseAndDescendants
$ haxe build.hxml --display src/Main.hx@75@usage@withDescendants
```

### CODE REVIEW

**Comment by Simn** [comment]
Shouldn't these be some sort of flags on the `"display/references"` request instead?

File: src/core/tFunctions.ml [hunk 1]
```diff
@@ -210,6 +210,9 @@ let rec is_parent csup c =
 		| None -> false
 		| Some (c,_) -> is_parent csup c
 
+(* for convenience *)
+let extends c csup = is_parent csup c
```
Comment (Simn) [inline src/core/tFunctions.ml]: This seems superfluous.

**Comment by nadako** [comment]
Is there performance difference between these methods? Because I'm wondering if we should just always return everything and let IDEs figure out what to show. E.g. I imagine it would be nice to only ask user whether they want override usages if there are any overrides in the first place (not that VSCode supports asking anyway though).

**Comment by Gama11** [comment]
Well, for that the IDE would still need some metadata for each reference regarding whether it's a base reference or not. I imagined it could work the other way around, with the IDE providing a parameter that indicates whether or not to include descendants (could be a vshaxe setting).

File: src/core/tFunctions.ml (hunk 1 above)
Comment (RealyUniqueName) [reply src/core/tFunctions.ml]: I was looking for a method to check if `c` extends `csup` for a few minutes. And then even started to write one. And only then I realized it could be "is_parent" instead of "extends" :)

File: src/context/display/findReferences.ml [hunk 2]
```diff
@@ -25,18 +26,74 @@ let find_references tctx com with_definition name pos kind =
...
+	match kind, com.display.dms_kind with
+	| SKField (cf,Some cl_path), DMUsage (_,find_descendants,find_base) when find_descendants || find_base ->
+		let collect() =
+			let c =
+				let rec loop = function
+					| [] -> raise Exit
+					| TClassDecl c :: _ when c.cl_path = cl_path -> c
+					| _ :: types -> loop types
+				in
+				loop com.types
```
Comment (Simn) [inline src/context/display/findReferences.ml:44]: Is there a particular reason we store the class' path and then dig it up here, instead of just storing the `tclass` reference itself like we usually do?

File: src/core/tFunctions.ml (hunk 1 above)
Comment (Simn) [reply src/core/tFunctions.ml]: Well, I actually prefer `extends` because you can infix-read it as `c extends csup`. But we shouldn't have both methods with different argument order.

**Comment by RealyUniqueName** [comment]
> Is there performance difference between these methods?

Yes. These methods perform as many  reference searches as  the amount of descendant methods in hierarchy. 

And now I realized this could be done in one search :) Will update later.

File: src/context/display/findReferences.ml (hunk 2 above)
Comment (RealyUniqueName) [reply src/context/display/findReferences.ml:44]: Yes. 
`tclass` instances are not physically equal [at this moment](https://github.com/RealyUniqueName/haxe/blob/1a5ae13/src/context/display/displayEmitter.ml#L105)  and at the moment of `find_references` invocation.
Maybe that's a bug, but I didn't investigate that.

File: src/core/tFunctions.ml (hunk 1 above)
Comment (RealyUniqueName) [reply src/core/tFunctions.ml]: Ok, I'll refactor `is_parent` in a separate commit to development.

**Comment by RealyUniqueName** [comment]
Changed to a parameter for `display/references` request instead of a separate request types.

**Comment by Gama11** [comment]
Hm, having three kinds confuses me a bit. What exactly is the "old behavior" right now / "normal"?

File: src/compiler/displayOutput.ml [hunk 3]
```diff
@@ -243,7 +243,13 @@ let handle_display_argument com file_pos pre_compilation did_something =
 				DMDefinition
 			| "usage" ->
 				Common.define com Define.NoCOpt;
-				DMUsage false
+				DMUsage (false,false,false)
+			| "usage@descendants" ->
```
Comment (Gama11) [inline src/compiler/displayOutput.ml]: I wonder if we should still add support for new features to the legacy protocol?

Comment (RealyUniqueName) [reply src/compiler/displayOutput.ml]: It's used for our display tests.

**Comment by RealyUniqueName** [comment]
> What exactly is the "old behavior" right now / "normal"?

It finds only "direct" references. No parent or overridden methods.

File: src/compiler/displayOutput.ml (hunk 3 above)
Comment (Gama11) [reply src/compiler/displayOutput.ml]: Sure, but I'm not sure it should be. :)

**Comment by Gama11** [comment]
Right... Not sure `Normal` is a good name / an enum is a good way to model this. In particular, I find this description a bit confusing:

>Find references to the base field **instead** of the requested one.
>Also finds references to all descendants of the base field.

So, not really "instead" if it finds references to the requested one after all? :)

I wonder if two boolean flags would work better?

```haxe
typedef FindReferencesParams = {
    var ?includeBase:Bool;
    var ?includeDescendants:Bool;
}
```

Though I'm not sure if we would even want to expose the second as a setting, I think IntelliJ only lets you chose between base fields or not?

**Comment by RealyUniqueName** [comment]
I've changed wording:
```haxe
/**
	Find only direct references to the requested symbol.
	Does not look for references to parent or overriding methods.
**/
var Direct = "direct";
/**
	Find references to the base field and all the overriding fields in the inheritance chain.
**/
var Base = "base";
/**
	Find references to the requested field and references to all
	descendants of the requested field.
**/
var Descendants = "descendants";
```


> I wonder if two boolean flags would work better?

I don't think so. There's actually no practical sence in this combination: 
```
{
  includeBase:true,
  includeDescendants:false
}
```

**Comment by Gama11** [comment]
Technically these are `Direct`, `WithBaseAndDescendants` and `WithDescendants` then. Not sure if that's too wordy though.

**Comment by RealyUniqueName** [comment]
Changed to `Direct`, `WithBaseAndDescendants` and `WithDescendants`.

**Comment by RealyUniqueName** [comment]
@Simn are you ok with this PR?

**Comment by Simn** [comment]
Yes, except for the `usage@withDescendants` stuff. Let's not extend the old display API, I really don't want to maintain that longer than necessary.

**Comment by Gama11** [comment]
I wonder if we should think ahead a bit and name it `ReferencesKind` instead of `FindReferencesKind`? Most likely we will need the exact same differentiation for an eventual `display/rename` method, right?

Or perhaps "Find References" makes sense for rename anyway too?

**Comment by nadako** [comment]
I'm still not 100% convinced that we need all these settings in practice, but if we do then we should be consistent and also make "find implementations" configurable like that. Because right now when searching for implementations of an interface we also get indirect implementations.

**Comment by RealyUniqueName** [comment]
> I wonder if we should think ahead a bit and name it ReferencesKind

That sounds more like a property for each reference found.

**Comment by RealyUniqueName** [comment]
> I'm still not 100% convinced that we need all these settings in practice

Based on my experience with a big Haxe project development in IDEA I'm 100% sure these settings are useful in practice :)
Well, except maybe `Direct` setting, which is the current behavior of "find references"

**Comment by RealyUniqueName** [comment]
HL failure is not related
